api/document/serializers/content.py

```diff
@@ -111,19 +127,34 @@ def get_content_type(self, instance: NestableAnnotation):
         return self.CONTENT_TYPE
 
     def get_inlines(self, instance: NestableAnnotation):
+        if self.IS_LEAF:
+            return []
         return NestedAnnotationSerializer(
             instance.children, context=self.context, many=True).data
 
     def get_text(self, instance: Annotation):
         return self.doc_node_text[instance.start:instance.end]
 
+    def to_internal_value(self, data: PrimitiveDict) -> PrimitiveDict:
+        inlines = [] if self.IS_LEAF else data.get('inlines', [])
+        result = super().to_internal_value(data)
+        result['content_type'] = self.CONTENT_TYPE
+        result['inlines'] = []
+        if inlines:
+            serializer = NestedAnnotationSerializer()
+            result['inlines'].extend([
+                serializer.to_internal_value(item) for item in inlines
+            ])
+        if self.IS_LEAF:
+            serializer = serializers.CharField(trim_whitespace=False)
+            result['text'] = serializer.to_internal_value(data.get('text'))
+        return result
```

Comment:
Overriding the serializer's `to_internal_value()` was more convenient than turning `CONTENT_TYPE`, `inlines`, and `text` into specialized fields, but we could do that instead if we want.